lib/csapi/message_pagination.h

```diff
@@ -25,29 +25,39 @@ class QUOTIENT_API GetRoomEventsJob : public BaseJob {
      * \param roomId
      *   The room to get events from.
      *
+     * \param dir
+     *   The direction to return events from. If this is set to `f`, events
+     *   will be returned in chronological order starting at `from`. If it
+     *   is set to `b`, events will be returned in *reverse* chronological
+     *   order, again starting at `from`.
+     *
      * \param from
      *   The token to start returning events from. This token can be obtained
-     *   from a `prev_batch` token returned for each room by the sync API,
-     *   or from a `start` or `end` token returned by a previous request
-     *   to this endpoint.
+     *   from a `prev_batch` or `next_batch` token returned by the `/sync`
+     * endpoint, or from an `end` token returned by a previous request to this
+     * endpoint.
      *
-     * \param dir
-     *   The direction to return events from.
+     *   This endpoint can also accept a value returned as a `start` token
+     *   by a previous request to this endpoint, though servers are not
+     *   required to support this. Clients should not rely on the behaviour.
+     *
+     *   If it is not provided, the homeserver shall return a list of messages
+     *   from the first or last (per the value of the `dir` parameter) visible
+     *   event in the room history for the requesting user.
      *
      * \param to
      *   The token to stop returning events at. This token can be obtained from
-     *   a `prev_batch` token returned for each room by the sync endpoint,
-     *   or from a `start` or `end` token returned by a previous request to
-     *   this endpoint.
+     *   a `prev_batch` or `next_batch` token returned by the `/sync` endpoint,
+     *   or from an `end` token returned by a previous request to this endpoint.
      *
      * \param limit
      *   The maximum number of events to return. Default: 10.
      *
      * \param filter
      *   A JSON RoomEventFilter to filter returned events with.
      */
-    explicit GetRoomEventsJob(const QString& roomId, const QString& from,
-                              const QString& dir, const QString& to = {},
+    explicit GetRoomEventsJob(const QString& roomId, const QString& dir,
```

Comment:
Yes it is. and the `Fix Room::getPreviousContent()` says so. Guess we need to start using [designated initializers](https://en.cppreference.com/w/cpp/language/aggregate_initialization#Designated_initializers) to catch those things. To enable them would take definining intermediate structures holding all the parameters, and jobs would accept those structures instead of individual parameters. An interesting piece to experiment with after 0.7.

Reply:
Also, in this particular case it would massively help if GTAD supported enumerations...